File: atlantic_report/atlantic_report/report/supplier_do_report___detail/supplier_do_report___detail.js
// Copyright (c) 2025
// Supplier DO Report - Detail

frappe.query_reports["Supplier DO Report - Detail"] = {

    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.month_start(),
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        },
        {
            fieldname: "supplier",
            label: __("Supplier"),
            fieldtype: "Link",
            options: "Supplier"
        }
    ],

    skip_total_row: 1,   // total row dihandle manual di PY
    onload(report) {

        const style = document.createElement("style");
        style.innerHTML = `

            /* =====================================================
               GLOBAL FONT 12PX
            ===================================================== */
            .dt-cell, .dt-cell__content, .dt-cell__text {
                font-size: 12px !important;
                line-height: 1.25 !important;
            }

            /* =====================================================
               HEADER ROW AUTO HEIGHT — ONLY HEADER
            ===================================================== */
            .dt-row-header {
                height: auto !important;
                min-height: auto !important;
                display: flex !important;
                align-items: flex-start !important;
            }

            .dt-row-header .dt-cell {
                height: auto !important;
                min-height: auto !important;
                display: flex !important;
                align-items: flex-start !important;
            }

            .dt-row-header .dt-cell__content {
                height: auto !important;
                min-height: auto !important;
                display: block !important;
                white-space: normal !important;
                line-height: 1.2 !important;
                overflow: visible !important;
            }

            /* =====================================================
               BODY ROW — DO NOT TOUCH HEADER OR FILTER
            ===================================================== */
            .dt-body .dt-row:not(.dt-row-header):not(.dt-row-filter) {
                height: 65px !important;
                min-height: 65px !important;
            }

            .dt-body 
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
            .dt-cell__wrapper,
            .dt-body 
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
            .dt-cell__content {
                height: 65px !important;
                min-height: 65px !important;
            }

            /* =====================================================
               FIX ITEM DISPLAY — ONLY BODY, NEVER HEADER/FILTER
            ===================================================== */
            .dt-row:not(.dt-row-header):not(.dt-row-filter) .dt-cell--item_display {
                overflow: visible !important;
            }

            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--item_display .dt-cell__wrapper {
                height: auto !important;
                overflow: visible !important;
                display: block !important;
                align-items: flex-start !important;
            }

            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--item_display .dt-cell__content {
                height: auto !important;
                overflow: visible !important;
                display: block !important;
                white-space: normal !important;
            }

            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--item_display .dt-cell__text {
                white-space: normal !important;
                display: block !important;
                overflow: visible !important;
            }

            /* =====================================================
               RIGHT ALIGN — PRICE & AMOUNT ONLY IN BODY ROW
            ===================================================== */
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--rate .dt-cell__content,
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--rate .dt-cell__text,
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--amount .dt-cell__content,
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--amount .dt-cell__text {
                text-align: right !important;
                justify-content: flex-end !important;
                width: 100% !important;
                display: flex !important;
            }
        `;
        document.head.appendChild(style);

        // ==================================================================
        // PATCH DATATABLE: OVERRIDE internal getRowHeight() → 65px
        // ==================================================================
        const patch_row_height = () => {
            const dt = frappe.query_report?.datatable;
            if (!dt) return;

            const H = 65;

            // set cell height globally
            dt.options.cellHeight = H;

            const rm = dt.rowmanager;
            if (!rm) return;

            // simpan original
            if (!rm.__orig_getRowHeight) {
                rm.__orig_getRowHeight = rm.getRowHeight?.bind(rm);
            }

            // override seluruh row height
            rm.getRowHeight = function(rowIndex) {
                return H;
            };

            rm.rowHeight = H;
            rm.row_height = H;

            try {
                rm.refreshRows(dt.datamanager.getRows());
            } catch(e) {
                dt.refresh();
            }
        };

        frappe.after_ajax(patch_row_height);
        setTimeout(patch_row_height, 350);
        setTimeout(patch_row_height, 900);
    },

    // ==================================================================
    // FORMATTER UTAMA
    // ==================================================================
    formatter(value, row, column, data, default_formatter) {

        // Jalankan default formatter (angka, html link, dsb)
        value = default_formatter(value, row, column, data);

        if (!data) {
            return value;
        }

        // ==================================================================
        // BASE STYLE: semua cell font 12px
        // ==================================================================
        let style = `
            font-size:12px;
            line-height:1.15;
            margin:0 !important;
            padding:0 !important;
            vertical-align:top !important;
            display:inline-block;
            width:100%;
        `;

        // ==================================================================
        // ALIGNMENT
        // ==================================================================
        if (["Currency", "Float", "Int"].includes(column.fieldtype) ||
            ["qty_info", "rate_info", "amount_info"].includes(column.fieldname)) {
            style += "text-align:right !important; white-space:nowrap;";
        }

        if (column.fieldname === "currency_info") {
            style += "text-align:center !important;";
        }

        // ==================================================================
        // BOLD RULES
        // ==================================================================

        // Row total → seluruh kolom bold
        if (data.is_total_row) {
            style += "font-weight:bold !important;";
        }

        // Supplier name → bold hanya kolom ini (bukan seluruh row)
        if (column.fieldname === "supplier_name") {
            style += "font-weight:bold !important;";
        }

        // ==================================================================
        // Column berisi HTML (gunakan div, bukan span)
        // ==================================================================
        const html_columns = [
            "pr_info", "supplier_name", "status",
            "item_info", "qty_info", "currency_info",
            "rate_info", "amount_info"
        ];

        if (html_columns.includes(column.fieldname)) {
            return `<div style="${style}">${value}</div>`;
        }

        // selain itu → span biasa
        return `<span style="${style}">${value}</span>`;
    }
};

File: atlantic_report/atlantic_report/report/supplier_do_report___detail/supplier_do_report___detail.py
# Copyright (c) 2025, Takwindo
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import formatdate, flt, add_months, today, fmt_money
import re

# ============================================================
# AGGRESSIVE TEXT SANITIZER
# ============================================================
def clean_text(text):
    """Bersihkan text dari newline, tab, spasi berlebih, dll."""
    if text is None:
        return "-"
    t = str(text)

    # Hapus newline, carriage return, tab
    t = t.replace("\n", " ").replace("\r", " ").replace("\t", " ")

    # Multi-space -> single space
    t = re.sub(r"\s{2,}", " ", t)

    # Hapus whitespace di antara tag HTML
    t = re.sub(r">\s+<", "><", t)

    t = t.strip()
    return t or "-"

# ============================================================
def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

# ============================================================
def get_columns():
    return [
        {
            "label": "PR NO / PO NO<br>PR DATE / PO DATE<br>RECEIVED BY",
            "fieldname": "pr_info",
            "fieldtype": "Data",
            "width": 230,
            "align": "left",
        },
        {
            "label": "SUPPLIER NAME",
            "fieldname": "supplier",
            "fieldtype": "Data",
            "width": 200,
        },
        {
            "label": "STATUS",
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": "ITEM CODE / BRAND<br>DESCRIPTION",
            "fieldname": "item_info",
            "fieldtype": "Data",
            "width": 300,
            "align": "left",
        },
        {
            "label": "QTY<br>UOM",
            "fieldname": "qty_info",
            "fieldtype": "Data",
            "width": 90,
            "align": "right",
        },
        {
            "label": "CCY",
            "fieldname": "currency_info",
            "fieldtype": "Data",
            "width": 50,
            "align": "center",
        },
        {
            "label": "PRICE",
            "fieldname": "rate_info",
            "fieldtype": "Data",
            "width": 110,
            "align": "right",
        },
        {
            "label": "TOTAL",
            "fieldname": "amount_info",
            "fieldtype": "Data",
            "width": 120,
            "align": "right",
        },
    ]

# ============================================================
def get_data(filters):
    if not filters:
        filters = frappe._dict()

    if not filters.get("from_date"):
        filters.from_date = add_months(today(), -1)
    if not filters.get("to_date"):
        filters.to_date = today()

    sql = """
        SELECT
            pr.name AS pr_no,
            pr.posting_date AS pr_date,
            pr.custom_received_by,
            pr.supplier,
            pr.status,
            pr.currency,
            pr.grand_total,

            pri.purchase_order AS po_no,
            po.transaction_date AS po_date,

            pri.item_code,
            pri.item_name,
            pri.description,
            pri.qty,
            pri.uom,
            pri.rate,
            pri.amount,

            i.brand

        FROM `tabPurchase Receipt` AS pr
        JOIN `tabPurchase Receipt Item` AS pri ON pri.parent = pr.name
        LEFT JOIN `tabPurchase Order` AS po ON po.name = pri.purchase_order
        LEFT JOIN `tabItem` AS i ON i.name = pri.item_code
        WHERE
            pr.docstatus = 1
            AND pr.posting_date BETWEEN %(from_date)s AND %(to_date)s
        ORDER BY
            pr.name ASC
    """

    raw_data = frappe.db.sql(sql, filters, as_dict=True)

    # Style basic (tanpa bold, font-size 10px)
    style_norm = "font-size:10px; color:#333;"

    processed = []
    last_pr = None

    totals_by_ccy = {}

    for row in raw_data:
        ccy = clean_text(row.currency)

        if ccy not in totals_by_ccy:
            totals_by_ccy[ccy] = {"grand_total": 0.0, "qty": 0.0}

        totals_by_ccy[ccy]["qty"] += flt(row.qty)

        # ========== CLEAN TEXT FIELD ==========
        pr_no_txt   = clean_text(row.pr_no)
        po_no_txt   = clean_text(row.po_no) if row.po_no else "-"
        rec_by_txt  = clean_text(row.custom_received_by) if row.custom_received_by else "-"
        supp_name   = clean_text(row.supplier)
        status_txt  = clean_text(row.status)
        item_code   = clean_text(row.item_code)
        item_name   = clean_text(row.item_name)
        brand_txt   = clean_text(row.brand) if row.brand else "-"
        # description fallback ke item_name kalau kosong
        desc_src    = row.description if row.description else item_name
        desc_txt    = clean_text(desc_src)

        # ========== ITEM INFO: 3 BARIS KONSISTEN ==========
        line1 = f'<span style="{style_norm}">{item_code} / {brand_txt}</span>'
        line2 = f'<span style="{style_norm}">{desc_txt}</span>'
        line3 = ''

        row["item_info"] = line1 + "<br>" + line2

        # ========== QTY INFO ==========
        qty_val = flt(row.qty)
        row["qty_info"] = (
            f'<span style="{style_norm}">{qty_val:,.2f}</span><br>'
            f'<span style="{style_norm}">{clean_text(row.uom)}</span>'
        )

        # ========== PRICE & AMOUNT ==========
        price = fmt_money(row.rate, currency=ccy)
        amt = fmt_money(row.amount, currency=ccy)

        row["rate_info"] = f'<span style="{style_norm}">{price}</span>'
        row["amount_info"] = f'<span style="{style_norm}">{amt}</span>'

        # ========== HEADER PER PR ==========
        if last_pr == row.pr_no:
            # Baris lanjutan dari PR yang sama -> kosongkan info header
            row["pr_info"] = ""
            row["supplier"] = ""
            row["status"] = ""
            row["currency_info"] = ""
        else:
            totals_by_ccy[ccy]["grand_total"] += flt(row.grand_total)

            pr_date_str = formatdate(row.pr_date, "dd-MMM-yy") if row.pr_date else "-"
            po_date_str = formatdate(row.po_date, "dd-MMM-yy") if row.po_date else "-"

            row["pr_info"] = (
                f'<span style="{style_norm}">{pr_no_txt}</span> / '
                f'<span style="{style_norm}">{po_no_txt}</span><br>'
                f'<span style="{style_norm}">{pr_date_str}</span> / '
                f'<span style="{style_norm}">{po_date_str}</span><br>'
                f'<span style="{style_norm}">Rec By: {rec_by_txt}</span>'
            )

            # supplier TIDAK dibold di sini (bold diatur di JS/HTML)
            row["supplier"] = f'<span style="{style_norm}">{supp_name}</span>'
            row["status"] = f'<span style="{style_norm}">{status_txt}</span>'
            row["currency_info"] = f'<span style="{style_norm}">{ccy}</span>'

            last_pr = row.pr_no

        processed.append(row)

    # =====================================================
    # TOTAL ROW PER CURRENCY
    # =====================================================
    if processed:
        for ccy, vals in totals_by_ccy.items():
            style_total = "font-size:10px; font-weight:bold; color:#333;"

            total_amt_html = (
                f'<span style="{style_total}">'
                f'{fmt_money(vals["grand_total"], currency=ccy)}'
                f'</span>'
            )
            total_qty_html = (
                f'<span style="{style_total}">{flt(vals["qty"]):,.2f}</span>'
            )

            label_html = (
                '<div style="text-align:right; font-size:10px; font-weight:bold;">'
                f'TOTAL ({ccy}) :'
                '</div>'
            )

            processed.append(
                {
                    "pr_info": label_html,
                    "supplier": "",
                    "status": "",
                    "item_info": "",
                    "qty_info": total_qty_html,
                    "currency_info": f'<span style="{style_total}">{ccy}</span>',
                    "rate_info": "",
                    "amount_info": total_amt_html,
                    "is_total_row": True,
                }
            )

    return processed
